File: project/sumin-onboarding/survey-service/application/src/main/java/com/innercircle/survey/application/service/SurveyResponseService.java
package com.innercircle.survey.application.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.innercircle.survey.common.dto.AnswerDto;
import com.innercircle.survey.common.dto.SurveyResponseDto;
import com.innercircle.survey.domain.entity.Question;
import com.innercircle.survey.domain.entity.Survey;
import com.innercircle.survey.domain.entity.SurveyResponse;
import com.innercircle.survey.domain.repository.SurveyRepository;
import com.innercircle.survey.domain.repository.SurveyResponseRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class SurveyResponseService {

    private final SurveyRepository surveyRepository;
    private final SurveyResponseRepository surveyResponseRepository;
    private final ObjectMapper objectMapper;

    public void submitResponse(SurveyResponseDto request) {
        // 1. 설문 존재 여부 확인
        Survey survey = surveyRepository.findById(request.getSurveyId())
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 설문입니다."));

        // 2. 질문 ID 유효성 검증
        Set<UUID> validQuestionIds = survey.getQuestions().stream()
                .map(Question::getId)
                .collect(Collectors.toSet());

        for (AnswerDto answer : request.getAnswers()) {
            UUID qid = UUID.fromString(answer.getQuestionId());
            if (!validQuestionIds.contains(qid)) {
                throw new IllegalArgumentException("유효하지 않은 질문 ID입니다: " + qid);
            }
        }

        // 3. 응답 및 질문 스냅샷 JSON 직렬화
        String answersJson = toJson(request.getAnswers());
        String snapshotJson = toJson(survey.getQuestions());

        // 4. 저장
        SurveyResponse response = new SurveyResponse(
                survey.getId(),
                answersJson,
                snapshotJson
        );
        surveyResponseRepository.save(response);
    }

    private String toJson(Object obj) {
        try {
            return objectMapper.writeValueAsString(obj);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("JSON 변환 실패", e);
        }
    }
}

File: project/kyungrok-onboarding/src/main/kotlin/kr/innercircle/onboarding/survey/service/SurveyResponseService.kt
package kr.innercircle.onboarding.survey.service

import kr.innercircle.onboarding.survey.domain.Survey
import kr.innercircle.onboarding.survey.domain.SurveyResponse
import kr.innercircle.onboarding.survey.dto.request.CreateSurveyResponseRequest
import kr.innercircle.onboarding.survey.repository.SurveyResponseRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * packageName : kr.innercircle.onboarding.survey.service
 * fileName    : SurveyResponseService
 * author      : ckr
 * date        : 25. 6. 21.
 * description :
 */

@Service
class SurveyResponseService(
    private val surveyResponseRepository: SurveyResponseRepository,
    private val surveyResponseAnswerService: SurveyResponseAnswerService
) {

    @Transactional
    fun createSurveyResponses(
        survey: Survey,
        createSurveyResponseRequest: CreateSurveyResponseRequest
    ): SurveyResponse {
        val surveyResponse = surveyResponseRepository.save(
            SurveyResponse.from(survey, createSurveyResponseRequest)
        )

        surveyResponseAnswerService.createSurveyResponseAnswers(surveyResponse, createSurveyResponseRequest.answers)
        return surveyResponse
    }
}